presto-spi/src/main/java/com/facebook/presto/spi/plan/SetOperationNode.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.plan;
+
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+@Immutable
+public abstract class SetOperationNode
+        extends PlanNode
+{
+    private final List<PlanNode> sources;
+    private final LinkedHashMap<VariableReferenceExpression, List<VariableReferenceExpression>> outputToInputs;
+    private final List<VariableReferenceExpression> outputVariables;
+
+    @JsonCreator
+    protected SetOperationNode(
+            @JsonProperty("id") PlanNodeId id,
+            @JsonProperty("sources") List<PlanNode> sources,
+            @JsonProperty("outputToInputs") LinkedHashMap<VariableReferenceExpression, List<VariableReferenceExpression>> outputToInputs)
```

Comment:
Does it make sense to require a separate `List<VariableReferenceExpression> outputs` and a `Map<VariableReferenceExpression, List<VariableReferenceExpression>> outputToInputs` to decouple the two requirements (ordered output vs map from output to inputs).